README.md
# scroll_cool_widget

|Vertical scroll|Horizontal scroll|
|---|---|
|![vertical](https://raw.githubusercontent.com/InnoFlutter/scroll_cool_widget/main/img/vertical.gif)|![horizontal](https://raw.githubusercontent.com/InnoFlutter/scroll_cool_widget/main/img/horizontal.gif)|

## Getting started

In your pubspec.yaml

```
dependencies:
  scroll_cool_widget: ^0.0.1
```

```dart
import 'package:scroll_cool_widget/scroll_cool_widget.dart';
```

## Usage

### With default settings (without back button)

```dart
ScrollCoolWidget(
  countOfCache: 10,
  height: (MediaQuery.of(context).size.height),
  width: (MediaQuery.of(context).size.width),
  
  createWidget: (index, width, height) {
    return Container(
      width: width,
      height: height,
      color: Color((Random().nextDouble() * 0xFFFFFF).toInt()).withOpacity(1.0),
      child: Center(
        child: Text('${index + 1}',
          style: const TextStyle(
            fontSize: 30,
          ),
        ),
      ),
    );
  },
)
```

### With custom settings (with back button)

```dart
ScrollCoolWidget(
  countOfCache: 10,
  height: (MediaQuery.of(context).size.height),
  width: (MediaQuery.of(context).size.width),
  scrollDirection: Axis.vertical,
  buttonPosition: ButtonPosition.topLeft,
  buttonColor: Colors.blueAccent,
  buttonIconColor: Colors.white,
  
  createWidget: (index, width, height) {
    return Container(
      width: width,
      height: height,
      color: Color((Random().nextDouble() * 0xFFFFFF).toInt()).withOpacity(1.0),
      child: Center(
        child: Text('${index + 1}',
          style: const TextStyle(
            fontSize: 30,
          ),
        ),
      ),
    );
  },
)
```

## Parameters

|Parameters|Description|
|---|---|
|`Widget Function(int index, double width, double height) createWidget`|Function for creating new widgets by index|
|`int countOfCache`|Count of widgets which will be loaded in advance|
|`double width`|Width of page|
|`double height`|Height of page|
|`Axis scrollDirection`|Direction of scrolling|
|`ButtonPosition buttonPosition`|Position of button for back|
|`Color buttonColor`|Color of button for back|
|`Color buttonIconColor`|Color of icon in button for back|

### ButtonPosition

Available values:

* `ButtonPosition.none` - default
* `ButtonPosition.topLeft`
* `ButtonPosition.topRight`
* `ButtonPosition.bottomLeft`
* `ButtonPosition.bottomRight`

## License

Here is using [MIT license](./LICENSE).

CHANGELOG.md
## 0.0.1

* Create widget

## 0.0.1+1

* Update links to gifs

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:scroll_cool_widget_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/scroll_cool_widget.dart
import 'package:flutter/material.dart';
import 'package:scroll_cool_widget/button_position.dart';

class ScrollCoolWidget extends StatefulWidget {
  final Widget Function(int index, double width, double height) createWidget;
  final int countOfCache;
  final double width;
  final double height;
  final Axis scrollDirection;
  final ButtonPosition buttonPosition;
  final Color buttonColor;
  final Color buttonIconColor;

  late double _topPosition;
  late double _leftPosition;

  ScrollCoolWidget({
    Key? key,
    required this.createWidget,
    required this.countOfCache,
    required this.width,
    required this.height,
    this.scrollDirection = Axis.vertical,
    this.buttonPosition = ButtonPosition.none,
    this.buttonColor = Colors.blueAccent,
    this.buttonIconColor = Colors.white,
  }) : super(key: key) {
    switch (buttonPosition) {
      case ButtonPosition.none:
        _topPosition = 0;
        _leftPosition = 0;
        break;
      case ButtonPosition.topLeft:
        _topPosition = 10;
        _leftPosition = 10;
        break;
      case ButtonPosition.topRight:
        _topPosition = 10;
        _leftPosition = width - 65;
        break;
      case ButtonPosition.bottomLeft:
        _topPosition = height - 145;
        _leftPosition = 10;
        break;
      case ButtonPosition.bottomRight:
        _topPosition = height - 145;
        _leftPosition = width - 65;
        break;
    }
  }

  @override
  State<ScrollCoolWidget> createState() => _ScrollCoolWidgetState();
}

class _ScrollCoolWidgetState extends State<ScrollCoolWidget> {
  bool wasInit = false;

  List<Widget> cache = [];
  int index = 0;
  int globalIndex = 0;

  final PageController _controller = PageController(initialPage: 0);

  @override
  Widget build(BuildContext context) {
    _updateWidgets();

    return Stack(
      children: [
        PageView(
          controller: _controller,
          scrollDirection: widget.scrollDirection,
          children: cache,
          onPageChanged: (index) {
            setState(() {
              globalIndex += (index - this.index);
              this.index = index;
            });
          },
        ),
        if (widget.buttonPosition != ButtonPosition.none)
          Positioned(
            top: widget._topPosition,
            left: widget._leftPosition,
            child: FloatingActionButton(
              backgroundColor: widget.buttonColor,
              onPressed: () {
                _controller.animateToPage(0,
                  duration: const Duration(seconds: 2),
                  curve: Curves.ease,
                );

                setState(() {
                  index = 0;
                  globalIndex = 0;
                });
              },
              child: Icon(widget.scrollDirection == Axis.vertical ? Icons.arrow_upward : Icons.arrow_back,
                color: widget.buttonIconColor,
              ),
            ),
          ),
      ]
    );
  }

  _updateWidgets() {
    if (!wasInit) {
      for (int i = 0; i <= widget.countOfCache + 1; i++) {
        cache.add(widget.createWidget(i, widget.height, widget.width));
      }
      wasInit = true;
    }

    _clearWidgets();
    _addWidgets();
  }

  _addWidgets() {
    if (cache.length - index < widget.countOfCache + 1) {
      setState(() {
        cache.add(widget.createWidget(globalIndex + widget.countOfCache, widget.width, widget.height));
      });
    }
  }

  _clearWidgets() {
    if (cache.length - index > widget.countOfCache + 1) {
      setState(() {
        cache.removeAt(cache.length - 1);
      });
    }
  }
}


android/src/main/kotlin/com/scroll_cool_widget/scroll_cool_widget/ScrollCoolWidgetPlugin.kt
package com.scroll_cool_widget.scroll_cool_widget

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** ScrollCoolWidgetPlugin */
class ScrollCoolWidgetPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "scroll_cool_widget")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "getPlatformVersion") {
      result.success("Android ${android.os.Build.VERSION.RELEASE}")
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}


example/lib/main.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:scroll_cool_widget/button_position.dart';
import 'package:scroll_cool_widget/scroll_cool_widget.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  const Home({Key? key}) : super(key: key);

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Plugin example app'),
      ),
      body: ScrollCoolWidget(
        countOfCache: 1,
        height: (MediaQuery.of(context).size.height),
        width: (MediaQuery.of(context).size.width),
        scrollDirection: Axis.vertical,
        buttonPosition: ButtonPosition.topLeft,
        buttonColor: Colors.blueAccent,
        buttonIconColor: Colors.white,
        createWidget: (index, width, height) {
          return Container(
            width: width,
            height: height,
            color: Color((Random().nextDouble() * 0xFFFFFF).toInt()).withOpacity(1.0),
            child: Center(
              child: Text('${index + 1}',
                style: const TextStyle(
                  fontSize: 30,
                ),
              ),
            ),
          );
        },
      )
    );
  }
}

